simple_history/manager.py

```diff
@@ -89,14 +90,38 @@ def _as_of_set(self, date):
         model = type(self.model().instance)  # a bit of a hack to get the model
         pk_attr = model._meta.pk.name
         queryset = self.get_queryset().filter(history_date__lte=date)
-        for original_pk in set(queryset.order_by().values_list(pk_attr, flat=True)):
-            changes = queryset.filter(**{pk_attr: original_pk})
-            last_change = changes.latest("history_date")
-            if changes.filter(
-                history_date=last_change.history_date, history_type="-"
-            ).exists():
-                continue
-            yield last_change.instance
+        # If using MySQL, need to get a list of IDs in memory and then use them for the
+        # second query.
+        # Does mean two loops through the DB to get the full set, but still a speed
+        # improvement.
+        backend = connection.vendor
+        if backend == "mysql":
+            history_ids = {}
+            for item in queryset.order_by("-history_date", "-pk"):
+                if getattr(item, pk_attr) not in history_ids:
+                    history_ids[getattr(item, pk_attr)] = item.pk
+            latest_historics = queryset.filter(history_id__in=history_ids.values())
+        elif backend == "postgresql":
+            latest_pk_attr_historic_ids = (
+                queryset.order_by(pk_attr, "-history_date", "-pk")
+                .distinct(pk_attr)
+                .values_list("pk", flat=True)
+            )
+            latest_historics = queryset.filter(
+                history_id__in=latest_pk_attr_historic_ids
+            )
+        else:
+            latest_pk_attr_historic_ids = (
+                queryset.filter(**{pk_attr: OuterRef(pk_attr)})
+                .order_by("-history_date", "-pk")
+                .values("pk")[:1]
+            )
+            latest_historics = queryset.filter(
+                history_id__in=Subquery(latest_pk_attr_historic_ids)
+            )
```

Comment:
Why are these two different cases? Postgres can handle subqueries

Reply:
It can, but the postgres specific one uses `distinct(pk_attr)`, which could potentially have optimizations from the DB engine itself over a subquery - I admittedly don't know enough about Postgres's internals to know if there IS a performance gain for distinct over a subquery to be fair!

Your call on whether to use that or just use a subquery here.